web/src/app/allocations-drawer/allocations-drawer.component.ts

```diff
@@ -0,0 +1,122 @@
+import { ColumnDef } from "@app/models/column-def.model";
+import { CommonUtil } from "@app/utils/common.util";
+
+@Injectable()
+@Component({
+  selector: "app-allocations-drawer",
+  templateUrl: "./allocations-drawer.component.html",
+  styleUrls: ["./allocations-drawer.component.scss"],
+})
+export class AllocationsDrawerComponent implements OnInit {
+  @ViewChild("matDrawer", { static: false }) matDrawer!: MatDrawer;
+  @ViewChild("allocationMatPaginator", { static: true }) allocPaginator!: MatPaginator;
+  @ViewChild("allocSort", { static: true }) allocSort!: MatSort;
+  @Input() allocDataSource!: MatTableDataSource<AllocationInfo & { expanded: boolean }>;
+  @Input() selectedRow!: AppInfo | null;
+  @Input() externalLogsBaseUrl!: string | null;
+  @Input() partitionSelected!: string;
+  @Input() leafQueueSelected!: string;
+
+  @Output() removeRowSelection = new EventEmitter<void>();
+
+  allocColumnDef: ColumnDef[] = [];
+  allocColumnIds: string[] = [];
+  selectedAllocationsRow: number = -1;
+
+  ngOnChanges(): void {
+    if (this.allocDataSource) {
+      this.allocDataSource.paginator = this.allocPaginator;
+      this.allocDataSource.sort = this.allocSort;
+    }
+    console.log("!", this.externalLogsBaseUrl);
+  }
+
+  ngOnInit(): void {
+    this.allocColumnDef = [
+      { colId: "displayName", colName: "Display Name", colWidth: 1 },
+      { colId: "allocationKey", colName: "Allocation Key", colWidth: 1 },
+      { colId: "nodeId", colName: "Node ID", colWidth: 1 },
+      {
+        colId: "log",
+        colName: "Log Link",
+        colWidth: 1,
+      },
+      {
+        colId: "resource",
+        colName: "Resource",
+        colFormatter: CommonUtil.resourceColumnFormatter,
+        colWidth: 1,
+      },
+      { colId: "priority", colName: "Priority", colWidth: 0.5 },
+    ];
+    this.allocColumnIds = this.allocColumnDef.map((col) => col.colId);
+  }
+
+  formatResources(colValue: string): string[] {
+    const arr: string[] = colValue.split("<br/>");
+    // Check if there are "cpu" or "Memory" elements in the array
+    const hasCpu = arr.some((item) => item.toLowerCase().includes("cpu"));
+    const hasMemory = arr.some((item) => item.toLowerCase().includes("memory"));
+    if (!hasCpu) {
+      arr.unshift("CPU: n/a");
+    }
+    if (!hasMemory) {
+      arr.unshift("Memory: n/a");
+    }
+
+    // Concatenate the two arrays, with "cpu" and "Memory" elements first
+    const cpuAndMemoryElements = arr.filter((item) => item.toLowerCase().includes("CPU") || item.toLowerCase().includes("Memory"));
+    const otherElements = arr.filter((item) => !item.toLowerCase().includes("CPU") && !item.toLowerCase().includes("Memory"));
+    const result = cpuAndMemoryElements.concat(otherElements);
+
+    return result;
+  }
+
+  isAllocDataSourceEmpty() {
+    return this.allocDataSource?.data && this.allocDataSource.data.length === 0;
+  }
+
+  allocationsDetailToggle(row: number) {
```

Comment:
Can we get a unit test for all this nested logic?

Reply:
Missing tests are added to the spec file.